pdg/src/builder.rs

```diff
@@ -102,42 +109,102 @@ impl EventKindExt for EventKind {
 ) {
     use EventKind::*;
     match *event_kind {
-        Alloc { ptr, .. } => {
+        Alloc { ptr, size, .. } => {
+            // TODO: check for overlap with existing allocations
+            mapping.size = Some(size);
             provenances.insert(ptr, mapping);
         }
-        CopyPtr(ptr) => {
-            // only insert if not already there
-            if let Err(..) = provenances.try_insert(ptr, mapping) {
-                log::warn!("0x{:x} already has a source", ptr);
-            }
-        }
-        Realloc { new_ptr, .. } => {
+        Realloc { new_ptr, size, .. } => {
+            mapping.size = Some(size);
             provenances.insert(new_ptr, mapping);
         }
-        Offset(_, _, new_ptr) => {
-            provenances.insert(new_ptr, mapping);
+        Free { ptr } if ptr != 0 => {
+            // TODO: handle the case where `ptr` falls inside an allocation
+            let old_prov = provenances.remove(&ptr);
+            if old_prov.is_none() {
+                log::warn!("Tried to free invalid pointer 0x{:x}", ptr);
+            }
         }
-        CopyRef => {
-            provenances.insert(metadata.destination.clone().unwrap().local.into(), mapping);
+        CopyPtr(ptr) |
+        Project(_, ptr) |
+        Offset(_, _, ptr) => {
+            // Check that the pointer falls inside an existing allocation
+            let need_insert = provenances.range(0..=ptr)
+                .next_back()
+                .map(|(start, pi)| {
+                    // If the existing pointer has a size, check it
+                    // Otherwise, insert ours if it's different
+                    let size = match pi.size {
+                        Some(size) => size,
+                        None => return ptr != *start
+                    };
+
+                    let end = start.saturating_add(size);
+                    if (*start..=end).contains(&ptr) {
+                        return false;
+                    }
+
+                    if matches!(*event_kind, CopyPtr(_)) {
+                        // A new pointer that we don't know where it came from
+                        // Insert it for now, we should decide later what to do
+                        //
+                        // Assume an unknown pointer points to a word-sized object
+                        log::warn!("Pointer of unknown origin 0x{:x}", ptr);
+                        return true;
+                    }
+
+                    log::warn!("0x{:x} outside of object bounds 0x{:x}-0x{:x}",
+                        ptr, start, end);
+
+                    false
+                }).unwrap_or(true);
+
+            if need_insert {
+                provenances.insert(ptr, mapping);
+            }
         }
-        AddrOfLocal(ptr, _) => {
+        CopyRef => unimplemented!(),
+        AddrOfLocal(ptr, _, size) => {
+            // TODO: is this a local from another function?
+            mapping.size = size.try_into().ok();
```

Comment:
Can this conversion ever fail?  If not, it seems like you could make the field `size: usize` instead of `Option<usize>` and avoid the `None` case above.

Reply:
The field would still need to be `size: Option<usize>` because we have other cases where we get a pointer without a known size. That comes from the `None` on line 268.

Some of this is happening because we get pointers from native libraries (e.g. glibc) which we have no control over. I think the best we can do right now is say "this is a pointer we know nothing about", and maybe add some clever heuristics later.

Re the conversion: I could do `Some(size.try_into().unwrap())`, that might be better for debugging?